### ISSUE TITLE
Closing the calendar fires both onOpen and onClose events. (MUI 5)

### ISSUE THREAD

**Issue by kenlyon** [mui/mui-x#12539]
### Steps to reproduce

Link to live example: https://v5.mui.com/x/react-date-pickers/date-picker/

```
<LocalizationProvider dateAdapter={AdapterDayjs}>
  <DatePicker
    label="Basic example"
    value={value}
    onChange={(newValue) => {
      setValue(newValue);
    }}
    onOpen={() => alert("onOpen")}
    onClose={() => alert("onClose")}
    renderInput={(params) => <TextField {...params} />}
  />
</LocalizationProvider>
```

Steps:
1. Add event handlers for `onOpen` and `onClose` to log or alert.
2. Run the example.
3. Click the calendar to open it. You should see that the `onOpen` event was fired.
4. Click the calendar again to close it.

Expected: Only the `onClose` event fired.

Actual: The `onOpen` event is fired, then `onClose`.

The same is true for `DateTimePicker` and `TimePicker` components.

PS: This bug is not present in MUI 6 or MUI 7 but I don't yet have plans to upgrade, so if it's possible to have a bug fix in 5.x that would be really useful.

### Current behavior

Closing the calendar fires both the `onOpen` and `onClose` events.

### Expected behavior

Closing the calendar should fire only the `onClose` event.

### Context

I'm trying to keep track of when the calendar is open to prevent updating other state in my component while it's open. Also, I wanted to use the `open` prop tied to a state value of my own, toggled via onOpen and onClose. The current behaviour means every time I close it, the `onOpen` sets my `open` to `true`.

### Your environment

<details>
  <summary><code>npx @mui/envinfo</code></summary>

```
  System:
    OS: Windows 10 10.0.19045
  Binaries:
    Node: 20.9.0 - C:\Program Files\nodejs\node.EXE
    npm: 10.1.0 - C:\Program Files\nodejs\npm.CMD  
    pnpm: Not Found
  Browsers:
    Chrome: Not Found
    Edge: Chromium (122.0.2365.92)
  npmPackages:
    @emotion/react: ^11.10.5 => 11.11.4 
    @emotion/styled: ^11.10.5 => 11.11.0 
    @mui/material: 5.14.19 => 5.14.19 
    @mui/system: 5.14.19 => 5.14.19 
    @mui/x-date-pickers-pro: ^5.0.0 => 5.0.20      
    @types/react: ^17.0.53 => 17.0.76 
    react: ^18.2.0 => 18.2.0 
    react-dom: ^18.2.0 => 18.2.0 
    typescript: ~5.3.2 => 5.3.3
```
</details>

Tested in:
* Firefox: `124.0.1`
* Chrome: `123.0.6312.59`

tsconfig:
```
{
    "compilerOptions": {
        "target": "esnext",
        "lib": ["esnext", "dom"],
        "allowJs": true,
        "skipLibCheck": true,
        "esModuleInterop": true,
        "allowSyntheticDefaultImports": true,
        "strict": true,
        "forceConsistentCasingInFileNames": true,
        "module": "esnext",
        "moduleResolution": "node",
        "noFallthroughCasesInSwitch": true,
        "noImplicitReturns": false,
        "noImplicitThis": true,
        "noImplicitAny": false,
        "resolveJsonModule": true,
        "isolatedModules": true,
        "jsx": "react-jsx",
        "noEmit": true
    },
    "include": ["src"]
}
```

**Search keywords**: onOpen onClose DatePicker TimePicker DateTimePicker

**Comment by LukasTy** [comment]
Hello @kenlyon 
I'm sorry to hear that you are experiencing this problem, but it does not seem like a critical issue.
As per [our policy](https://mui.com/x/introduction/support/#supported-versions), the v5 is out of the support window (and even while v6 was stable, the v5 support policy encapsulates critical and security issues only).

Besides, it seems quite easy to avoid this problem by using [a controlled open state](https://stackblitz.com/edit/react-kzdnf4?file=demo.tsx). 😉 

I'm closing this issue, please consider upgrading in the future to benefit from various bugfixes, improvements and new components. 👍

**Comment by kenlyon** [comment]
@LukasTy Thanks for the info. I had thought it'd be likely we need to upgrade.

Just so you know, though, using a controlled open state didn't help. In fact, the first place I saw the issue, I was setting `open` based on a state property. It might vary, but in my case my code in the `onOpen` handler was causing some kind of re-render that was preventing the subsequent `onClose` from being received. The symptom was that the calendar was stuck open.

I mention that just in case anyone else was experiencing the same and wondering if it was just them. Either way, I'll plan to upgrade.

**Comment by LukasTy** [comment]
> I was setting `open` based on a state property. It might vary, but in my case my code in the `onOpen` handler was causing some kind of re-render that was preventing the subsequent `onClose` from being received. The symptom was that the calendar was stuck open.

@kenlyon It seems strange. Would you be able to share a minimal reproduction case?
It could be a combination of different implementation factors that lead to such behavior. 🙈

**Comment by kenlyon** [comment]
@LukasTy I spent quite a bit of time on this today but couldn't solve the mystery. I distilled my component down to the bare minimum and tested it in a fresh vite+react environment. Everything worked correctly. Although clicking the calendar button when it was open _did_ still fire the onOpen event, it also fired onClose on the same render cycle and so the net result was the `open` value was set to `false` and the calendar disappeared.

I copied the stripped down version of the component back into my original application and the problem persists. Clicking the calendar button when it's already open only fires onOpen, even though all that's happening there is `setIsOpen(true)`. No `onClose` event is fired and there's no subsequent render.

It must be something higher up the component hierarchy and it'll take me too long to unravel all of that, especially as I have a workaround.

We're upgrading to 7.x very soon anyway, so this problem won't affect us.
